parameter_flash_storage.c

```diff
@@ -0,0 +1,203 @@
+#include <string.h>
+#include "parameter_flash_storage.h"
+#include "parameter_flash_storage_private.h"
+#include "flash.h"
+#include "parameter/parameter_msgpack.h"
+#include "cmp/cmp.h"
+#include "cmp_mem_access/cmp_mem_access.h"
+#include "crc/crc32.h"
+
+/* We cannot use a CRC start value of 0 because CRC(0, 0xffffffff) = 0xffffffff
+ * which makes empty flash pages valid. */
+#define CRC_INITIAL_VALUE 0xdeadbeef
+
+
+static size_t cmp_flash_writer(struct cmp_ctx_s *ctx, const void *data, size_t len)
+{
+    cmp_mem_access_t *mem = (cmp_mem_access_t*)ctx->buf;
+    if (mem->index + len <= mem->size) {
+        flash_write(&mem->buf[mem->index], data, len);
+        mem->index += len;
+        return len;
+    } else {
+        return 0;
+    }
+}
+
+void parameter_flash_storage_erase(void *dst)
+{
+    flash_unlock();
+    flash_sector_erase(dst);
+    flash_lock();
+}
+
+static void err_mark_false(void *arg, const char *id, const char *err)
+{
+    (void) id;
+    (void) err;
+
+    bool *b = (bool *)arg;
+    *b = false;
+}
+
+void parameter_flash_storage_save(void *dst, size_t dst_len, parameter_namespace_t *ns)
+{
+    cmp_ctx_t cmp;
+    cmp_mem_access_t mem;
+    uint32_t len;
+    bool success = true;
+
+    void *orig_dst = dst;
+
+    flash_unlock();
+
+    /* If there is no valid block, erase flash just to start from a pristine
+     * state. */
+    if (parameter_flash_storage_block_find_last_used(dst) == NULL) {
+        flash_sector_erase(dst);
+    }
+
+    /* Find first available flash block. */
+    dst = parameter_flash_storage_block_find_first_free(dst);
+
+    len = dst_len - (dst - orig_dst);
+
+    /* If the destination is too small to fit even the header, erase the block. */
+    if (len <= PARAMETER_FLASH_STORAGE_HEADER_SIZE) {
+        flash_sector_erase(orig_dst);
+        dst = orig_dst;
+        len = dst_len;
```

Comment:
[question] Erase but not abort?

Reply:
Yes. The idea is that all config writes are appended at the end of the config sector, which is erase only when there is not enough room left. The write then starts over at the beginning. The idea behind this is to provide flash wear levelling by spreading the writes.

Reply:
works for me :)